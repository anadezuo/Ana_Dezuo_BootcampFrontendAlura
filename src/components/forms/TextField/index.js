import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../../foundation/Text';

const InputWrapper = styled.div`
  margin-bottom: 17px;
`;

const Input = styled(Text)`
  width: 100%;
  border: 1px solid ${({ theme }) => theme.colors.secondary.main.color};
  padding: 12px 16px;
  outline: 0;
  border-radius: ${({ theme }) => theme.borderRadius};
  background-color: inherit;
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph',
};

export default function TextField({
  placeholder,
  name,
  onChange,
  value,
  type,
  ...props
}) {
  return (
    <InputWrapper>
      <Input
        placeholder={placeholder}
        type={type}
        name={name}
        onChange={onChange}
        value={value}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
      />
    </InputWrapper>
  );
}

TextField.propTypes = {
  placeholder: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};
