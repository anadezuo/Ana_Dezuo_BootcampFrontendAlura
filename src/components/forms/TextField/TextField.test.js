import React from 'react';
import user from '@testing-library/user-event';
import { render, screen } from '../../../infra/test/testUtils';

import TextField from './index';

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        type="p"
        placeholder="Seu nome"
        value="Ju"
        onChange={() => {}}
        name="name"
      />
    );
    const textField = screen.getByPlaceholderText(/nome/i);
    expect(textField).toMatchSnapshot();
  });

  describe('when field is valid', () => {
    describe('and user is typing', () => {
      test('the value must be updated', () => {
        const onChangeMock = jest.fn();
        render(
          <TextField
            type="p"
            placeholder="Seu nome"
            value=""
            onChange={onChangeMock}
            name="name"
            isTouched
          />
        );
        const inputNome = screen.getByPlaceholderText(/nome/i);
        user.type(inputNome, 'Ana Teste');
        expect(onChangeMock).toHaveBeenCalledTimes(9);
      });
    });
  });

  describe('when field is invalid', () => {
    test('displays the respective error message', () => {
      render(
        <TextField
          type="p"
          placeholder="email@dominio.com.br"
          value="ana@teste.com"
          onChange={() => {}}
          name="email"
          isTouched
          error='"Email" é obrigatório'
        />
      );

      const inputEmail = screen.getByPlaceholderText(/email/i);
      expect(inputEmail).toHaveValue('ana@teste.com');
      expect(screen.getByRole('alert')).toHaveTextContent(
        '"Email" é obrigatório'
      );
      expect(inputEmail).toMatchSnapshot();
    });
  });
});
