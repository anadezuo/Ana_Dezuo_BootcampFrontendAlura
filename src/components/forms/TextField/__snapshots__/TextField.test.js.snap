// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  font-size: 20px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  border: 1px solid #212121;
  padding: 12px 16px;
  outline: 0;
  border-radius: 5px;
  background-color: inherit;
}

<input
  class="c0 c1"
  href=""
  name="name"
  placeholder="Seu nome"
  type="p"
  value="Ju"
/>
`;

exports[`<TextField /> when field is invalid displays the respective error message 1`] = `
.c0 {
  font-size: 20px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  border: 1px solid #212121;
  padding: 12px 16px;
  outline: 0;
  border-radius: 5px;
  background-color: inherit;
  border-color: #dc3545;
}

.c1 + span {
  color: #dc3545;
  font-size: 11px;
}

<input
  class="c0 c1"
  href=""
  name="email"
  placeholder="email@dominio.com.br"
  type="p"
  value="ana@teste.com"
/>
`;
